src/middlewared/middlewared/restful.py

```diff
@@ -529,30 +551,109 @@ async def do(self, http_method, req, resp, **kwargs):
+                        'errno': errno.EINVAL,
+                    }, indent=True)
+                    return resp
+
+                has_request_body = True
+                try:
+                    request_body = json.loads(await part.read())
+                except ValueError as e:
+                    resp.set_status(400)
+                    resp.headers['Content-type'] = 'application/json'
+                    resp.text = json.dumps({
+                        'message': f'`data` json parse error: {e}',
+                        'errno': errno.EINVAL,
+                    }, indent=True)
+                    return resp
+
+                filepart = await reader.next()
+                if not filepart or filepart.name != "file":
+                    resp.set_status(400)
+                    resp.headers['Content-type'] = 'application/json'
+                    resp.text = json.dumps({
+                        'message': ('The method accepts multipart requests with two parts (`data` and `file`). '
+                                    '`file` not found.'),
+                        'errno': errno.EINVAL,
+                    }, indent=True)
+                    return resp
+
+                upload_pipe = self.middleware.pipe()
+            else:
+                if method['check_pipes']:
+                    resp.set_status(400)
+                    resp.headers['Content-type'] = 'application/json'
+                    resp.text = json.dumps({
+                        'message': 'This method accepts only multipart requests.',
+                        'errno': errno.EINVAL,
+                    }, indent=True)
+                    return resp
+                else:
+                    if await req.text():
+                        has_request_body = True
+                        request_body = await req.json()
         else:
+            if await req.text():
+                has_request_body = True
+                request_body = await req.json()
+
+        download_pipe = None
+        if method['downloadable']:
+            if req.query.get('download', '1') == '1':
+                download_pipe = self.middleware.pipe()
+            else:
+                if method['check_pipes']:
+                    resp.set_status(400)
+                    resp.headers['Content-type'] = 'application/json'
+                    resp.text = json.dumps({
+                        'message': 'JSON response is not supported for this method.',
+                        'errno': errno.EINVAL,
+                    }, indent=True)
+                    return resp
+
+        if upload_pipe and download_pipe:
+            method_kwargs['pipes'] = Pipes(input=upload_pipe, output=download_pipe)
+        elif upload_pipe:
+            method_kwargs['pipes'] = Pipes(input=upload_pipe)
+        elif download_pipe:
+            method_kwargs['pipes'] = Pipes(output=download_pipe)
+
+        if True:
```

Comment:
Is this required ?

Reply:
I keep it for smaller and more readable diff. I'll remove this and unindent the code once this PR once it is reviewed and aproved